pkg/hydrator/asset_vulnerability_hydrator.go

```diff
@@ -0,0 +1,154 @@
+package hydrator
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/asecurityteam/nexpose-vuln-hydrator/pkg/domain"
+)
+
+// AssetVulnerabilityHydrator implements the domain.Hydrator interface for hydrating
+// an asset's vulnerabilities
+type AssetVulnerabilityHydrator struct {
+	Client domain.NexposeClient
+}
+
+// HydrateVulnerabilities takes a domain.Asset and fetches its vulnerabilities to be further
+// hydrated concurrently
+func (a *AssetVulnerabilityHydrator) HydrateVulnerabilities(asset domain.Asset) (domain.AssetVulnerabilityDetails, error) {
+	// For an asset - we first get the list of vulnerabilities for a given asset ID
+	// For each vulnerability, we can fetch these concurrently, as seen on lines 35-41
+	// For each vulnerability, we can fetch the details and solutions concurrently, as seen on lines 75-102
+	// The solutions response gives a list of solution IDs, we can again fetch the solution details concurrently as seeon on lines 134-153
+
+	// 1. fetch vulnerabilities to be hydrated - this contains contains the vulnerability ID (for further hydration)
+	// as well as the results (port and protocol information of finding)
+	vulnerabilities, err := a.Client.GetAssetVulnerabilities(asset.ID)
+	if err != nil {
+		return domain.AssetVulnerabilityDetails{}, err
+	}
+
+	vulnDetailsChan := make(chan domain.VulnerabilityDetails, len(vulnerabilities))
+	errChan := make(chan error)
+
+	var wg sync.WaitGroup
+	wg.Add(len(vulnerabilities))
+	for _, vuln := range vulnerabilities {
+		// 2. fetch vulnerabilities concurrently, placing results and errors on channels
+		go a.hydrateVulnerability(&wg, vuln, vulnDetailsChan, errChan)
+	}
+	wg.Wait()
+
+	// 17. Again not sure how to capture all errors or if we want to
```

Comment:
you will need to close the channel in order to range over them.

Alternatively, if you want to error out on a non-zero amount of errors, and want to return the first error you find, another way to do this would be:

```
for _ = range vulnerabilities {
    select {
    case err := <- errChan:
        return domain.AssetVulnerabilityDetails{}, err
    case details := <- vulnDetailsChan:
        vulnDetails = append(vulnDetails, detail)
    }
}
```

Note that this method would not require a waitgroup

Reply:
This is a good call-out, I think I was maybe thinking of errors incorrectly. I initially thought that no solutions might be a valid error case, but I think that the only errors we'd want to surface would be genuine API errors.

Although, what happens in the scenario where we would have multiple errors? I think the first would be caught and returned to the caller, but all other errors would be lost, right? Not sure how much we care about that, but it's an interesting trade-off.